Allow MVT layer filtering and add layer name to properties

jesusbotella opened the pull request:
According to what we talked about in #674, MVT loader lacks two nice-to-have features that are implemented in this pull request.

These features are:
- **MVT layer filtering**
   Developer can filter the layers that the loader will parse by using `layers` option and providing an array containing the names of the layers to parse. If any of those layers are not present, the loader will raise an error.

- **Adding layer name to GeoJSON properties**
Right now, you cannot know which layer a particular feature belongs to. It is very handy when trying to split or filter features by layers, so it's nice to have it. 
  - By default, layer's name will be added to `layer` property within GeoJSON properties.
  - If there's a `layer` property already present and the developer provides a fallback name in `layerProperty`, the loader will add that property containing layer's name.
  - If no fallback is provided, layer's name will not be added to properties' object.

The PR lacks documentation, and I need to complete some tests, but I open it for discussion.

ibgreen commented on modules/mvt/src/lib/parse-mvt.js:
```diff
@@ -40,9 +48,17 @@ function getDecodedFeature(feature, options = {}) {
+
+  // Add layer name to GeoJSON properties
+  const hasLayerProperty = decodedFeature.properties.hasOwnProperty('layer');
+  const layerPropertyName = hasLayerProperty ? options.layerProperty : 'layer';
```
Concern: My thought is that if `options.mvt.layerProperty` is provided, we always use that. This code seems to be using the user specified option name only if the `layer` prop is already defined on the tile. It will be hard for the application to know what prop to looks at.

ibgreen commented on modules/mvt/src/lib/transform-to-local-range.js:
```diff
@@ -67,7 +67,7 @@ function generateJSON(type, feature, coordinates) {
       coordinates
     },
-    properties: feature.properties
+    properties: feature.properties || {}
```
I see you caught it here, ignore my previous comment.

ibgreen commented on modules/mvt/test/mvt-loader.spec.js:
```diff
@@ -149,3 +151,21 @@ test('Should raise an error when coordinates param is wgs84 and tileIndex is mis
   t.end();
 });
+
+test('Should add layer name to custom property when layer property is already present', async t => {});
```
Assume these are WIP

jesusbotella commented on modules/mvt/src/lib/parse-mvt.js:
```diff
@@ -14,24 +14,32 @@ export default function parseMVT(input, options) {
+  selectedLayers.forEach(layerName => {
     const vectorTileLayer = tile.layers[layerName];
 
+    if (!vectorTileLayer) {
```
Yes, I thought it would be good to know if any of the layer names you had are not present. But throwing an error is maybe too much!

kylebarron commented on modules/mvt/src/lib/parse-mvt.js:
```diff
@@ -54,10 +54,11 @@ function getDecodedFeature(feature, options = {}, layerName) {
+  if (options.layerProperty) {
+    decodedFeature.properties[options.layerProperty] = options.layerName;
+  } else if (!isLayerPropertyPresent) {
+    decodedFeature.properties.layer = options.layerName;
```
I think we decided `layerName` was going to be the default. https://github.com/uber-web/loaders.gl/issues/674#issuecomment-597344966

jesusbotella replied:
Yes, sorry about that, I am just changing it now.

jesusbotella commented:
Just pushed a commit, but do not review it yet, as I misunderstood one comment 😅.

jesusbotella commented:
I think I addressed all review comments, but let me know if you want me to change something else.

I made it to not override the original property in case of collision but @ibgreen's comment made me change my mind:
> the layer should always be added, either in layer or the layerProperty. If you really want to be able to prevent this prop from being added, why not just make layerPropertydefault to layer and support layerProperty: null if the user don't want it to be added at all.

`layerName` property may not be as common as `layer`, so I think we're good. But let me know if you want me to change it to avoid overriding any original parameters.

Pessimistress commented on modules/mvt/src/lib/parse-mvt.js:
```diff
@@ -14,19 +14,27 @@ export default function parseMVT(input, options) {
       const vectorTileFeature = vectorTileLayer.feature(i);
 
-      const decodedFeature = getDecodedFeature(vectorTileFeature, mvtOptions);
+      const decodedFeature = getDecodedFeature(vectorTileFeature, {...loaderOptions, layerName});
```
Perf: construct this object outside of the for loop

jesusbotella replied:
Oh sorry! :( Didn't notice of this!

ibgreen commented on modules/mvt/src/mvt-loader.js:
```diff
@@ -16,7 +16,8 @@ export const MVTWorkerLoader = {
       tileIndex: {},
-      workerUrl: `https://unpkg.com/@loaders.gl/mvt@${VERSION}/dist/mvt-loader.worker.js`
+      workerUrl: `https://unpkg.com/@loaders.gl/mvt@${VERSION}/dist/mvt-loader.worker.js`,
+      layerProperty: 'layerName'
```
Nit; Order of props seems a little random. Maybe alphabetical order (perhaps with the cross-loader `workerUrl`) prop at the top or bottom would look a little cleaner?

ibgreen commented on modules/mvt/docs/api-reference/mapbox-vector-tile-loader.md:
```diff
@@ -83,10 +83,12 @@ const geoJSONfeatures = await load(url, MVTLoader);
+| Option            | Type                                       | Default      | Description                                                                                                                                                                                                                                                                 |
+| ----------------- | ------------------------------------------ | ------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| mvt.coordinates   | String                                     | `local`      | When set to `wgs84`, parser will return a flat array of GeoJSON representations of the features with coordinates decoded from provided tile index. When set to `local`, parser will return a flat array of GeoJSON objects with local coordinates decoded from tile origin. |
+| mvt.layerProperty | String                                     | `layerName`  | The layer name is added to `feature.properties[layerProperty]` (first adds a `feature.property` object is created if it did not already exist). If set to `null`, it won't add layer name to any property.                                                                  |
```
`... it won't add layer name to any property` => `...a layer name property will not be added.`

ibgreen commented on modules/mvt/docs/api-reference/mapbox-vector-tile-loader.md:
```diff
@@ -83,10 +83,12 @@ const geoJSONfeatures = await load(url, MVTLoader);
+| ----------------- | ------------------------------------------ | ------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| mvt.coordinates   | String                                     | `local`      | When set to `wgs84`, parser will return a flat array of GeoJSON representations of the features with coordinates decoded from provided tile index. When set to `local`, parser will return a flat array of GeoJSON objects with local coordinates decoded from tile origin. |
+| mvt.layerProperty | String                                     | `layerName`  | The layer name is added to `feature.properties[layerProperty]` (first adds a `feature.property` object is created if it did not already exist). If set to `null`, it won't add layer name to any property.                                                                  |
+| mvt.layers        | String[]                                   | _No default_ | Optional list of layer names. If supplied, only features belonging to the named layers will be included in the output. If not supplied, features from all layers are returned.                                                                                              |
```
Nit: I think it makes sense to specify a default so that the app can contain code that conditionally sets this prop without if statements.  Perhaps `null` or `false`?

ibgreen commented on modules/mvt/docs/api-reference/mapbox-vector-tile-loader.md:
```diff
@@ -83,10 +83,12 @@ const geoJSONfeatures = await load(url, MVTLoader);
+| mvt.coordinates   | String                                     | `local`      | When set to `wgs84`, parser will return a flat array of GeoJSON representations of the features with coordinates decoded from provided tile index. When set to `local`, parser will return a flat array of GeoJSON objects with local coordinates decoded from tile origin. |
+| mvt.layerProperty | String                                     | `layerName`  | The layer name is added to `feature.properties[layerProperty]` (first adds a `feature.property` object is created if it did not already exist). If set to `null`, it won't add layer name to any property.                                                                  |
+| mvt.layers        | String[]                                   | _No default_ | Optional list of layer names. If supplied, only features belonging to the named layers will be included in the output. If not supplied, features from all layers are returned.                                                                                              |
+| mvt.tileIndex     | Object ({x: number, y: number, z: number}) | _No default_ | Mandatory with `wgs84` coordinates option. An object containing tile index values (x, y, z) to decode WGS84 coordinates.                                                                                                                                                    |
```
default is `null` or `false`?

ibgreen commented on modules/mvt/src/mvt-loader.js:
```diff
@@ -15,9 +15,9 @@ export const MVTWorkerLoader = {
   options: {
     mvt: {
       coordinates: 'local',
+      layerProperty: 'layerName',
```
Add layers default.

kylebarron commented:
I could write up a small docs PR. 

It might also be helpful to note in the docs that when working in vector tile local coordinates, the origin is in the top-left, not the bottom-left. I gather that's normal for some image formats, but as a relative newcomer I was confused for a while.
